spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableStrategy.scala

```diff
@@ -0,0 +1,267 @@
+  }
+
+  override def updateOptimizeStats(
+      optimizeStats: OptimizeStats,
+      removedFiles: Seq[RemoveFile],
+      bins: Seq[(Map[String, String], Seq[AddFile])]): Unit = {}
+}
+
+/** Implements ZOrder strategy */
+case class ZOrderOptimizeTableStrategy(
+    override val sparkSession: SparkSession,
+    zOrderColumns: Seq[String]) extends OptimizeTableStrategy {
+
+  assert(zOrderColumns.nonEmpty)
+
+  override val optimizeTableMode: OptimizeTableMode.Value = OptimizeTableMode.ZORDER
+
+  override val curve: String = "zorder"
+
+  // For ZORDER, set maxBinSize the maximal LONG value to have single BIN for each partition.
+  override val maxBinSize: Long = Long.MaxValue
+
+  override def updateOptimizeStats(
+      optimizeStats: OptimizeStats,
+      removedFiles: Seq[RemoveFile],
+      bins: Seq[(Map[String, String], Seq[AddFile])]): Unit = {
+    val inputFileStats =
+      ZOrderFileStats(removedFiles.size, removedFiles.map(_.size.getOrElse(0L)).sum)
+    optimizeStats.zOrderStats = Some(ZOrderStats(
+      strategyName = "all", // means process all files in a partition
+      inputCubeFiles = ZOrderFileStats(0, 0),
+      inputOtherFiles = inputFileStats,
+      inputNumCubes = 0,
+      mergedFiles = inputFileStats,
+      // There will one z-cube for each partition
+      numOutputCubes = optimizeStats.numPartitionsOptimized))
+  }
+}
+
+/** Implements clustering strategy for clustered tables */
+case class ClusteringOptimizeTableStrategy(
+    override val sparkSession: SparkSession,
+    clusteringColumns: Seq[String]) extends OptimizeTableStrategy {
+
+  override val optimizeTableMode: OptimizeTableMode.Value = OptimizeTableMode.CLUSTERING
+
+  override val curve: String = "hilbert"
+
+  /**
+   * In clustering, the bin size corresponds to a ZCube size that can be adjusted through
+   * configurations.
+   */
+  override val maxBinSize: Long = {
+    Math.max(
+      sparkSession.sessionState.conf.getConf(DELTA_OPTIMIZE_CLUSTERING_TARGET_CUBE_SIZE),
+      sparkSession.sessionState.conf.getConf(DELTA_OPTIMIZE_CLUSTERING_MIN_CUBE_SIZE))
+  }
+
+  // Prepare files for binpack.
+  override def prepareFilesPerPartition(inputFiles: Seq[AddFile]): Seq[AddFile] = {
+    // Un-clustered files don't have a ZCUBE_ID, and are sorted before clustered files (None is
+    // considered the smallest element). We also don't consider partitionValues because
+    // clustered tables should always be unpartitioned.
+    applyMinZCube(inputFiles.sortBy(_.tag(AddFile.Tags.ZCUBE_ID)))
+  }
+
+  // Upon a new ZCube, allocate a [[ZCubeInfo]] with a new ZCUBE ID.
+  override def initNewBin: OptimizeTableStrategy.BinInfo = {
+    OptimizeTableStrategy.ZCubeBinInfo(ZCubeInfo(clusteringColumns))
+  }
+
+  override def tagAddFile(file: AddFile, binInfo: OptimizeTableStrategy.BinInfo): AddFile = {
+    val taggedFile = super.tagAddFile(file, binInfo)
+    val zCubeInfo = binInfo.asInstanceOf[OptimizeTableStrategy.ZCubeBinInfo].zCubeInfo
```

Comment:
btw, can `initNewBin` instantiate the value internally, so that the user doesn't need to pass the `binInfo` here?